File: src/main/java/org/jcodec/containers/mp4/boxes/MdtaBox.java
package org.jcodec.containers.mp4.boxes;

import java.nio.ByteBuffer;

import org.jcodec.common.io.NIOUtils;
import org.jcodec.containers.mp4.boxes.Box.AtomField;
import org.jcodec.platform.Platform;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class MdtaBox extends Box {

    private static final String FOURCC = "mdta";
    private String key;

    public MdtaBox(Header header) {
        super(header);
    }

    public static MdtaBox createMdtaBox(String key) {
        MdtaBox box = new MdtaBox(Header.createHeader(FOURCC, 0));
        box.key = key;
        return box;
    }

    @Override
    public void parse(ByteBuffer buf) {
        key = Platform.stringFromBytes(NIOUtils.toArray(NIOUtils.readBuf(buf)));
    }

    @AtomField(idx=0)
    public String getKey() {
        return key;
    }

    @Override
    protected void doWrite(ByteBuffer out) {
        out.put(key.getBytes());
    }

    @Override
    public int estimateSize() {
        return key.getBytes().length;
    }
    
    public static String fourcc() {
        return FOURCC;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/MdtaBox.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.common.io.NIOUtils
import org.jcodec.platform.Platform
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
class MdtaBox(header: Header) : Box(header) {
    var key: String? = null
        private set

    override fun parse(buf: ByteBuffer) {
        key = Platform.stringFromBytes(NIOUtils.toArray(NIOUtils.readBuf(buf)))
    }

    override fun doWrite(out: ByteBuffer) {
        out.put(key!!.toByteArray())
    }

    override fun estimateSize(): Int {
        return key!!.toByteArray().size
    }

    companion object {
        private const val FOURCC = "mdta"
        @JvmStatic
        fun createMdtaBox(key: String?): MdtaBox {
            val box = MdtaBox(Header.createHeader(FOURCC, 0))
            box.key = key
            return box
        }

        @JvmStatic
        fun fourcc(): String {
            return FOURCC
        }
    }
}